### ISSUE TITLE
My little roadmap

### ISSUE THREAD

**Issue by agnat** [korfuri/client_cpp#13]
I'm thinking about giving the prometheus client some love. I think it makes sense to have a ticket to coordinate my work. Here is my tentative roadmap.

I think I'll start with a couple of smaller changes:
- Get rid of C-style memory managment of `MetricFamily`
- Implement binary output

Then I'd like to start looking at the API and would like to propose a few changes. First, I'd like to hide the compile time parameters of the metric types by using `auto`. So instead of

``` c++
Counter<2> pizzas_served("pizzas_served", "number of pizzas served", {"topping", "size"});
```

I'd like to write

``` c++
auto pizzas_served = makeCounter("pizzas_served", "number of pizzas served", 
                                 labels("topping", "size");
```

Using `auto` would permit us to put more compile time configuration into each metric without burdening the user with complex types. Then we could do things like:

``` c++
using DISABLED = prometheus::metric_options::disabled;


auto pizzas_served = makeCounter("pizzas_served", "number of pizzas served", 
                                 labels("topping", "size"), DISABLED());
```

The resulting counter would not self-register. All calls to the counter would remain valid but are turned into a `nop`. That sort of thing...

Using `labels(...)`, a variadic template function instead of a raw initializer list helps to capture the list size at compile time.

The call site might look like this:

``` c++
  pizzas_served.labels("parma", "humongous").inc();
```

Capturing the argument count here we could check it against the initial label count of the metric. We could provide a "strict metric policy" that won't compile unless all labels are provided. That is it only accepts points in data-space and not sub-spaces.

``` c++
using strict = prometheus::metric_options::requires_all_lables;

auto pizzas_served = makeCounter("pizzas_served", "number of pizzas served", 
                                 labels("topping", "size"), strict());

pizzas_served.labels("parma", "humongous").inc(); // compiles
pizzas_served.labels("funghi").inc();             // does not
```

Other interesting policies might be `metric_options::single_threaded` which skips the `std::atomic<>` and uses a plain old value, &c.

So, I guess I'm proposing a [policy based design](https://en.wikipedia.org/wiki/Policy-based_design) for metric types, which I believe will give us the flexibility to do really cool things (at compile time) while keeping the API clean and lean.

Let me know what you think.

**Comment by korfuri** [comment]
What do you mean by implementing binary output? `ICollector::collect` already gives protos objects ready to be sent to prometheus :) Do you mean adding that to the gnu_libmicrohttpd integration?

I like the policy-based design idea a lot, and this is actually closer to what I originally had in mind for the cpp client.

Nit: require_all_labels should probably be the default, wdyt?

**Comment by agnat** [comment]
> What do you mean by implementing binary output? ICollector::collect already gives protos objects ready to be sent to prometheus :)

Not quite. Protobuf objects don't contain framing information. See [Streaming Multiple Messages](https://developers.google.com/protocol-buffers/docs/techniques#streaming). That's why the server actually consumes a format called "protobuf delimited". Each protobuf object is preceded by its size encoded as a [varint](https://developers.google.com/protocol-buffers/docs/encoding#varints). I already have it working and it's just one simple function. But I think it's worth including since it took me quite a while to find the details...

> [...] and this is actually closer to what I originally had in mind for the cpp client.

Great! That's what I thought...

> Nit: require_all_labels should probably be the default, wdyt?

That's what I thought too but TBH I'm not sure yet. I just don't have enough operational experience with prometheus yet. We should take a look at the go client and see what it does...

**Comment by korfuri** [comment]
This sounds good overall. I'm travelling right now and for the next 3 weeks I will have limited time for this project, but I'll do my best to drop by once in a while to review changes :)
